test/test_js_utils.js

```diff
@@ -131,3 +131,104 @@ describe('blocks are given IDs and can be looked up', () => {
   })
 
 })
+
+describe('blocks return proper this.columns', () => {
+
+  beforeEach(() => {
+    TidyBlocksManager.reset()
+  })
+
+  it('groupBy returns additional _group_ column', (done) => {
+    const pipeline = [
+      makeBlock( // because the pipeline has to start with a data block
+        'data_single',
+        {}),
+      makeBlock(
+        'transform_groupBy',
+        {COLUMN: 'first'})
+    ]
+    const env = evalCode(pipeline)
+    assert(env.table[0].hasOwnProperty('_group_')) // check for _group_ column
+    done()
+  })
+
+  it('ungroup removes _group_ column', (done) => {
+    const pipeline = [
+      makeBlock( // because the pipeline has to start with a data block
+        'data_single',
+        {}),
+      makeBlock(
+        'transform_groupBy', // add the _group_ column
+        {COLUMN: 'first'}),
+      makeBlock(
+        'transform_ungroup', // remove _group_ column
+        {})
+    ]
+    const env = evalCode(pipeline)
+    assert(env.table[0].hasOwnProperty('_group_') === false) // check _group_ column removed
+    done()
+  })
+
+  // FIXME evaluating to falsey when true
+  it('sort this.columns unchanged', (done) => {
+    // original dataframe
+    const pipeline_iris = [
+      makeBlock(
+      'data_iris',
+      {})
+    ]
+    // sorted dataframe
+    const pipeline_sort = [
+      makeBlock(
+        'data_iris',
+        {}),
+      makeBlock(
+        'transform_sort',
+        {MULTIPLE_COLUMNS: 'Sepal_Length',
+         DESCENDING: 'false'})
+    ]
+    const env_iris = evalCode(pipeline_iris)
+    const env_sort = evalCode(pipeline_sort)
+    // console.log(Object.keys(env_iris.table[0]))
+    // console.log(Object.keys(env_sort.table[0]))
+    // compare pipeline columns
+    assert(Object.keys(env_iris.table[0]) == Object.keys(env_sort.table[0]))
```

Comment:
Ah - arrays in JavaScript are only equal if they're the same object - `==` doesn't recurse into the array to check values (because how deep should it go)? Use https://nodejs.org/api/assert.html#assert_assert_deepstrictequal_actual_expected_message instead

Reply:
Also, creating two pipelines is cool, but how about:

```
const pipeline = [
    makeBlock(
    'data_iris',
    {})
]
const env_iris = evalCode(pipeline)
pipeline.push(makeBlock(
        'transform_sort',
        {MULTIPLE_COLUMNS: 'Sepal_Length',
         DESCENDING: 'false'}))
const env_sort = evalCode(pipeline)
```
and then do your asserts? Saying `const x = [...some array]` means that `x` always points to the same array (i.e., the binding of `x` to an array is constant) but the array itself is still mutable.